---
title: "Processing Lorenzi gene expression data"
author: "Gitta Ekaputeri, Anne-Christin Hauschild"
date: "29/08/2022" 
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(stringr)
library(miRBaseConverter)
library(reshape2)
setwd("/data/RNA/lorenziData/")
```

### Read counts data

Load counts file
```{r}
oriCountsFile <- read.table(file = './GSE138734_RNAAtlas_totalRNA_counts_filtered.txt')
countsFile <- oriCountsFile[,-c(2)] # we don't need the second column (genome_coordinates)
countsFile <- janitor::row_to_names(countsFile, row_number=1)
head(countsFile)
```

### Combine the replicates

Get the meta information

```{r}
meta <- read.table("./meta_celltype_tissue.txt", header = T, sep = "\t") # meta consists only tissues and cell type, no cell lines
meta$RNAAtlas <- str_replace_all(meta$X.Sample_characteristics_ch1.1, "sample id \\[deprecated\\]: ", "") # RNAAtlas is the sample ID here
```


Create Disease, Tissue, and Group (Tissue_Disease) columns

```{r}
meta$Tissue <- str_replace_all(meta$X.Sample_source_name_ch1, " tissue", "")
meta$Tissue <- str_replace_all(meta$Tissue, "[^A-Za-z0-9]+", "\\.")
meta$Tissue <- str_replace_all(meta$Tissue, "-", ".")
meta$Tissue <- str_replace_all(meta$Tissue, " ", ".")
meta$Tissue <- tolower(meta$Tissue)
meta$Tissue <- str_replace_all(meta$Tissue, "cb.cd34.lin.", "cb.cd34.lin")
meta$Disease <- "" #all normal
meta$Group <- apply(meta[,c(ncol(meta)-1,ncol(meta))], 1, function(x) paste(x[1], x[2], sep="_"))
meta$Group <- str_replace_all(meta$Group, "_", "")
meta$Group <- str_replace(meta$Group, "\\.donor\\.[0-9]+", "") #Remove the `.donor.[0-9]` suffix that exists in some of the samples.
```


Grouping: Combine replicates (samples that have the same type)

```{r}
uniqueTissues <- unique(meta$Group)

# Remove columns in countsFile that does not exist in meta (that means the sample type is not tissue and not cell type)
names(countsFile) %in% meta$RNAAtlas
countsFile <- countsFile[,-(2:94),drop=FALSE]

# Remove RNAAtlas in meta that does not exist in countsFile
meta$RNAAtlas %in% names(countsFile)
meta <- meta[(meta$RNAAtlas %in% names(countsFile)),]


# Now all data in countsFile are tissues or cell type and all samples in meta exist in countsFile

countsFile[,2:ncol(countsFile)] = lapply(countsFile[,2:ncol(countsFile)], FUN = function(y){as.numeric(y)})
combinedData <- lapply(uniqueTissues, function(x) {
  columnsToFetch <- meta$RNAAtlas[which(meta$Group == x)] #combine data with the same RNAAtlas into 1 group (Because here, RNA Atlas is Sample ID)
  df <- as.data.frame(countsFile[,columnsToFetch])
  if(ncol(df) == 1) {
    names(df) = x
  }
  return(rowMeans(df))
})

names(combinedData) <- uniqueTissues
combinedData <- do.call(cbind, combinedData)
combinedData <- cbind(countsFile[,1], combinedData)
combinedData <- data.frame(combinedData)
names(combinedData)[names(combinedData) == 'V1'] <- 'Gene'
```


### Ensure the terms are standardized
Ensure that all the group names appear in the ontology or the corrections file.

Get the ontology and correction files

```{r}
ont <- read.table("C:/Users/gitta/Documents/MirDIP_counts/ontology.txt", header = F, sep = "\t")
corr <- read.table("C:/Users/gitta/Documents/MirDIP_counts/corrections.txt", header = T, sep = "\t")
```

Aim to have 0 as a result, meaning all terms are either already in the ontology or the correction file. Or else: update correction file or ontology file, or both.

```{r}
trms <- unique(unlist(strsplit(names(combinedData)[2:ncol(combinedData)], "_"))) # Splits the composite terms that contain both tissue+disease
if (length(which(trms %in% union(corr[,1], unique(unlist(ont[,c(1,3)]))))) != 0){
  trms[-which(trms %in% union(corr[,1], unique(unlist(ont[,c(1,3)]))))] # Which terms aren't in the ontology or the corrected terms
} else {
  trms
}
```
Remove granulocytes because it's too granular for us
```{r}
combinedData2 <- combinedData
drop <- c("cd10.granulocytes","cd14.cd15.granulocytes")
combinedData2 = combinedData2[,!(names(combinedData2) %in% drop)]
```


Correct current terms that need to be corrected.
```{r}
colnames(combinedData) <- sapply(colnames(combinedData), function(z) {
  y <- strsplit(z, "_")[[1]]
  retVal <- sapply(y, function(x) {
    if (x %in% corr$currentTerm) {
      return(corr$correctedTerm[match(x, corr$currentTerm)])
    } else {
      return(x)
    }
  })
  return(paste(retVal, collapse = "_"))
})
```


### Convert to long format
Add the Canonical column and Source column

```{r}
data <- combinedData # all tissues have to exist in ontology or correction files
data$Source <- "Lorenzi"
data <- data[,c(1,ncol(data),2:(ncol(data)-1))] # Set columns alignment
head(data)
```

Convert to long format

```{r}
data_long <- melt(data, id.vars=c("Gene", "Source"))
```

Binarization
```{r}
#data_long$Binary <- sapply(data_long$value, function(x) if (x == 0) return(0) else return(1))
#names(data_long) <- c("Gene", "Source", "Tissue", "Scale", "Binary")
#head(data_long)

# Remove non-binary value, keep only binary value.
data_long$Binary <- sapply(data_long$value, function(x) if (x == 0) return(0) else return(1))
data_long_final <- subset(data_long, select = -c(value) )
names(data_long_final) <- c("Gene", "Source", "Tissue", "Binary")
head(data_long_final)
```

### Write to file

```{r}
write.table(data_long_final, "./lorenzi_longData.txt", sep = "\t", row.names = F, col.names = T, quote = F)
write.table(unique(data_long_final[,1]), "./lorenzi_genes.txt", sep = "\t", row.names = F, col.names = T, quote = F)
writeLines(as.character(unique(data_long_final$Tissue)), "./lorenzi_tissues.txt")
```
